### PR TITLE
Logic Classes and Sets

### PR DESCRIPTION
**Pull request by ShikharJ**

### CODE REVIEW

File: symengine/lib/symengine_wrapper.pyx [hunk 1]
```diff
@@ -843,6 +885,160 @@ cdef class Constant(Basic):
...
+
+
+class And(Boolean):
+
+    def __new__(cls, *args):
+        cdef symengine.set_boolean s
+        cdef Boolean e_
+        for e in args:
+            e_ = sympify(e)
+            s.insert(e_.thisptr)
```
Comment (ShikharJ) [inline symengine/lib/symengine_wrapper.pyx]: @isuruf Why isn't `insert` identified as a part of `set_boolean`?

Comment (ShikharJ) [inline symengine/lib/symengine_wrapper.pyx]: Ping @isuruf.

Comment (isuruf) [inline symengine/lib/symengine_wrapper.pyx]: What do you mean by that?

Comment (ShikharJ) [inline symengine/lib/symengine_wrapper.pyx]: The builds fail because of the following error:
```
 Cythonizing symengine_wrapper.pyx
  
  Error compiling Cython file:
  ------------------------------------------------------------
  ...
  def logical_and(*args):
      cdef symengine.set_boolean s
      cdef Boolean e_
      for e in args:
          e_ = sympify(e)
          s.insert(e_.thisptr)
          ^
  ------------------------------------------------------------
  
  C:\projects\symengine-py-l1jmr\symengine\lib\symengine_wrapper.pyx:3043:9: Object of type 'set_boolean' has no attribute 'insert'
```

Comment (isuruf) [inline symengine/lib/symengine_wrapper.pyx]: Since it says it is not defined, you have to define it in the pxd file

**Comment by ShikharJ** [comment]
@isuruf I'm out of ideas on fixing the current build failure. Can you please take a look, and suggest what can be done?

**Comment by isuruf** [comment]
Here is the relevant line, https://github.com/symengine/symengine.py/blob/b409536eb9029b274d2956ec14a7954a0089aa3a/symengine/lib/symengine.pxd#L43
Similar line, https://github.com/symengine/symengine.py/blob/b409536eb9029b274d2956ec14a7954a0089aa3a/symengine/lib/symengine.pxd#L54

**Comment by ShikharJ** [comment]
@isuruf Those changes have already been applied. I was referring to the current build failure (the output was too big to post here), which you can view in the `Appveyor` builds.

File: symengine/lib/symengine.pxd [hunk 2]
```diff
@@ -203,7 +206,7 @@ cdef extern from "<symengine/basic.h>" namespace "SymEngine":
     ctypedef map[RCP[Integer], unsigned] map_integer_uint "SymEngine::map_integer_uint"
     cdef struct RCPIntegerKeyLess
     cdef struct RCPBasicKeyLess
-    ctypedef set[RCP[const_Basic], RCPBasicKeyLess] set_basic "SymEngine::set_basic"
+    ctypedef set[RCP[Basic], RCPBasicKeyLess] set_basic "SymEngine::set_basic"
```
Comment (isuruf) [inline symengine/lib/symengine.pxd]: Why this change? Might be the issue

Comment (ShikharJ) [inline symengine/lib/symengine.pxd]: I was getting this error without the above change:
```
Cythonizing symengine_wrapper.pyx
  
  Error compiling Cython file:
  ------------------------------------------------------------
  ...
  def finiteset(*args):
      cdef symengine.set_basic s
      cdef Basic e_
      for e in args:
          e_ = sympify(e)
          s.insert(e_.thisptr)
                    ^
  ------------------------------------------------------------
  
  C:\projects\symengine-py\symengine\lib\symengine_wrapper.pyx:3922:19: Cannot assign type 'RCP[const Basic]' to 'RCP[const_Basic]'
```

**Comment by ShikharJ** [comment]
@isuruf Any idea on how this can be fixed?

**Comment by ShikharJ** [comment]
@isuruf Can you help me in getting this sorted out?

File: symengine/lib/symengine.pxd (hunk 2 above)
Comment (isuruf) [inline symengine/lib/symengine.pxd]: Cast `e_.thisptr` to `RCP[const_Basic]` instead of the above change.

File: symengine/lib/symengine_wrapper.pyx [hunk 3]
```diff
@@ -3886,5 +4176,102 @@ def ccode(expr):
...
+def universalset():
+    return c2py(<RCP[const symengine.Basic]>(symengine.universalset()))
+
+
+def finiteset(*args):
+    cdef symengine.set_basic s
+    cdef Basic e_
+    for e in args:
+        e_ = sympify(e)
+        s.insert(<RCP[symengine.const_Basic]>(e_.thisptr))
```
Comment (ShikharJ) [inline symengine/lib/symengine_wrapper.pyx]: @isuruf This is still failing.

**Comment by ShikharJ** [comment]
Ping @isuruf.

File: symengine/lib/symengine_wrapper.pyx (hunk 3 above)
Comment (isuruf) [inline symengine/lib/symengine_wrapper.pyx]: The failure now is not because of that. Open `symengine_wrapper.cpp` and see which line from `symengine_wrapper.pyx` is causing the problem. You can find the line by scrolling above and looking for the first comment.

**Comment by ShikharJ** [comment]
@isuruf Can you help me in fixing the errors?

File: symengine/lib/symengine_wrapper.pyx (hunk 3 above)
Comment (isuruf) [inline symengine/lib/symengine_wrapper.pyx]: @ShikharJ, what did you get?

Comment (ShikharJ) [inline symengine/lib/symengine_wrapper.pyx]: I was referring to the errors in the current log. This is fixed. The issues are only arising while running tests.

File: symengine/lib/symengine_wrapper.pyx [hunk 4]
```diff
@@ -182,6 +182,34 @@ cdef c2py(RCP[const symengine.Basic] o):
         r = Function.__new__(conjugate)
     elif (symengine.is_a_PyNumber(deref(o))):
         r = PyNumber.__new__(PyNumber)
+    elif (symengine.is_a_Piecewise(deref(o))):
+        r = Piecewise.__new__(Piecewise)
```
Comment (isuruf) [inline symengine/lib/symengine_wrapper.pyx]: Use `Basic.__new__(Piecewise)` or otherwise you get a infinite cycle.

**Comment by ShikharJ** [comment]
This is ready for a review @isuruf.

File: symengine/lib/symengine_wrapper.pyx [hunk 5]
```diff
@@ -2409,6 +2515,144 @@ class Subs(Basic):
...
+    def __new__(self, *args):
+        return piecewise(*args)
+
+    def _sympy_(self):
+        import sympy
+        a = self.args
+        l = []
+        for i in range(0, len(a), 2):
+            l.append((a[i]._sympy_(), a[i + 1]._sympy_()))
+        return sympy.Piecewise(*l)
```
Comment (isuruf) [inline symengine/lib/symengine_wrapper.pyx]: This is not tested.

File: symengine/tests/test_solve.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
+from symengine.utilities import raises
+from symengine.lib.symengine_wrapper import (Interval, EmptySet, FiniteSet,
+											 I, oo, solve, Eq, Symbol)
```
Comment (isuruf) [inline symengine/tests/test_solve.py]: Can you fix the spacing here?

File: symengine/tests/test_sets.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
+from symengine.utilities import raises
+from symengine.lib.symengine_wrapper import (Interval, EmptySet, UniversalSet, FiniteSet,
+											 Union, Complement, ImageSet, ConditionSet,
+											 And, Or, oo, Symbol, true, Ge, Eq, Gt)
```
Comment (isuruf) [inline symengine/tests/test_sets.py]: Fix formatting here.

File: symengine/tests/test_sets.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    assert Interval(-oo, -oo) == EmptySet()
+    assert isinstance(Interval(1, 1), FiniteSet)
+
+    assert Interval(1, 0) == EmptySet()
+    assert Interval(1, 1, False, True) == EmptySet()
+    assert Interval(1, 1, True, False) == EmptySet()
+    assert Interval(1, 1, True, True) == EmptySet()
+    assert Interval(1, 2).union(Interval(2, 3)) == Interval(1, 3)
+
+
```
Comment (isuruf) [inline symengine/tests/test_sets.py]: Remove whitespace here

**Comment by ShikharJ** [comment]
Updated @isuruf.

**Comment by ShikharJ** [comment]
Thanks for the help @isuruf.
